### DISCUSSION TITLE
EACCES Permission issue when updating package-lock.json from docker on Windows

### DISCUSSION THREAD

**Discussion by nzacca** [renovatebot/renovate#15065] [Request Help] [answered]
Hello!

We have been having some difficulty getting Renovate to update the package-lock file in our repo and I was hoping someone here might be able to shed some light on what we're missing here. Everything else is working perfectly:
* PR is created
* package.json is updated properly

## Issue

We are receiving the following error in the logs when it attempts to update the `package-lock.json` file:

<details>
<summary>Error in Logs</summary>
<p>

```bash
DEBUG: lock file error (repository=redacted, branch=renovate/ramda-0.x)
       "err": {
         "killed": false,
         "code": 243,
         "signal": null,
         "cmd": "docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/bitbucket/redacted\":\"/tmp/renovate/repos/bitbucket/redacted\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/tmp/renovate/repos/bitbucket/redacted\" docker.io/renovate/node:16.13.2 bash -l -c \"install-tool npm 8.3.0 && hash -d npm 2>/dev/null || true && npm install --no-audit --ignore-scripts\"",
         "stdout": "
Installing legacy tool npm v8.3.0
added 1 package in 2s
8.3.0
",
         "stderr": "
npm notice
npm notice New minor version of npm available! 8.1.2 -> 8.6.0
npm notice Changelog: <https://**redacted**@8.6.0` to update!
npm notice 
npm ERR! code EACCES
npm ERR! syscall open
npm ERR! path /tmp/renovate/repos/bitbucket/redacted/package-lock.json
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, open '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'
npm ERR!  [Error: EACCES: permission denied, open '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
",
         "message": "Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/bitbucket/redacted\":\"/tmp/renovate/repos/bitbucket/redacted\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/tmp/renovate/repos/bitbucket/redacted\" docker.io/renovate/node:16.13.2 bash -l -c \"install-tool npm 8.3.0 && hash -d npm 2>/dev/null || true && npm install --no-audit --ignore-scripts\"
npm notice 
npm notice New minor version of npm available! 8.1.2 -> 8.6.0
npm notice Changelog: <https://**redacted**@8.6.0` to update!
npm notice 
npm ERR! code EACCES
npm ERR! syscall open
npm ERR! path /tmp/renovate/repos/bitbucket/redacted/package-lock.json
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, open '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'
npm ERR!  [Error: EACCES: permission denied, open '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
",
         "stack": "Error: Command failed: docker run --rm --name=renovate_node --label=renovate_child -v \"/tmp/renovate/repos/bitbucket/redacted\":\"/tmp/renovate/repos/bitbucket/redacted\" -v \"/tmp/renovate/cache\":\"/tmp/renovate/cache\" -e NPM_CONFIG_CACHE -e npm_config_store -w \"/tmp/renovate/repos/bitbucket/redacted\" docker.io/renovate/node:16.13.2 bash -l -c \"install-tool npm 8.3.0 && hash -d npm 2>/dev/null || true && npm install --no-audit --ignore-scripts\"
npm notice 
npm notice New minor version of npm available! 8.1.2 -> 8.6.0
npm notice Changelog: <https://**redacted**@8.6.0` to update!
npm notice 
npm ERR! code EACCES
npm ERR! syscall open
npm ERR! path /tmp/renovate/repos/bitbucket/redacted/package-lock.json
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, open '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'
npm ERR!  [Error: EACCES: permission denied, open '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'open',
npm ERR!   path: '/tmp/renovate/repos/bitbucket/redacted/package-lock.json'
npm ERR! }
npm ERR! 
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR! 
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:

    at ChildProcess.exithandler (node:child_process:399:12)
    at ChildProcess.emit (node:events:526:28)
    at ChildProcess.emit (node:domain:475:12)
    at maybeClose (node:internal/child_process:1092:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)"
       },
       "type": "npm"
DEBUG: No updated lock files in branch (repository=redacted, branch=renovate/ramda-0.x)
```

</p>
</details>

<details>
<summary>Combined Config</summary>
<p>

```bash
DEBUG: Combined config
       "config": {
         "platform": "bitbucket",
         "baseBranches": ["redacted"],
         "extends": ["config:base"],
         "repositories": ["redacted"],
         "lockFileMaintenance": {"enabled": true},
         "includePaths": ["package.json"],
         "packageRules": ["snipped for brevity"],
         "username": "***********",
         "password": "***********",
         "hostRules": [{"matchHost": "github.com", "token": "***********"}],
         "requireConfig": false,
         "binarySource": "docker",
         "prHourlyLimit": 2
       }
```

</p>
</details>

## Our Environment

We build the 32.17 slim image and copy the config directly to the image instead of passing it on the command line.

We are running the following command from a **Windows Host**:

`docker run --rm -it -v /c/docker/tmp:/tmp -v //var/run/docker.sock:/var/run/docker.sock -e LOG_LEVEL=debug renovate --pr-hourly-limit 2`

Note the difference from the sample for the `tmp` volume - we were unable to mount `/tmp`. We have also attempted to create a docker volume and mount that instead but that also results in the same issue.

Also note the difference for the docker socket. We had to place a second forward slash for the child docker command to run properly. This was sourced from: https://stackoverflow.com/a/41005007/12493424

Things we have tried to resolve the issue:
* Attempted to set the `dockerUser` and `--user` to `1000:0` or `root` as per [this thread](https://github.com/renovatebot/renovate/discussions/13961)
* Attempted to create a volume for the tmp folder: `docker volume create tmp` and mount that instead `-v tmp:/tmp`
* Attempted to build the slim image and create a tmp folder within the image and permission it for the ubuntu user
* Given Read/Write permissions to `Everyone` on the `c:\docker\tmp` folder on the Windows host.
* Attempted to run with `--binary-source=install` which does not produce an error but the package lock file does not get updated.

Any insight would be greatly appreciated.

**Comment by viceice** [comment, marked as answer]
binarySource=docker isn't supported on windows container. it only works for Linux containers, so you need WSL or something like that.

if you're using Linux containers on windows, you need to mount the tmp volume and docker socket from the docker Linux VM instead from the windows host.

what are the logs for binarySource=install?

please provide some more details about your docker runtime on windows.

**Reply by nzacca** [reply]
@viceice 

Thank you so much for the response. Please forgive my ignorance - I'm not an experienced docker user as yet but I'm learning as best as I can.

We are using Linux containers on Windows via WSL. We're using Docker Desktop 4.6.1 (76265). I was running the command from the Windows side (via powershell) which was working except for updating the package-lock. After switching to running the command from Linux as you suggested, the package-lock updated successfully. We did have to remove the existing branches before we could see the update to the lock file but that's just a small detail. Thank you for that.

I'm curious, would you happen to know what prevents the child docker command from working when we initially run from the Windows host? If I understand what's happening, is it correct to say that the inner docker command is attempting to mount the installation folder from the container to the child instance but when the child instance attempts to write to that folder, it runs into a permission error? I'm assuming that, for some reason, the permissions of the mounted volume in the child are not correctly set/the user doesn't have permissions from the child context, is that accurate?

For completeness, we did try running with binarySource=install but that only worked when executing the command from Linux and not the Windows host. Unfortunately, the wrong version of node is installed on the image for us so we're opting to use the docker source instead.

Very much appreciate your time in helping us. Thank you again.

**Reply by viceice** [reply]
You can run those docker images from windows, but you can't use a windows folder as shared renovate dir.
For the slim image the host and container path need to match, as renovate otherwise doesn't know how tou mount it to the sidecars. Thats way those fail, as the mount simply doesn't exist, so docker creates a root owned folder. That causes the access denied to the renovaste user (uid=1000)

**Reply by viceice** [reply]
I usally run somethink like this from windows:

```pwsh
> docker run --rm `
    -e LOG_LEVEL=debug `
    -e CONFIG=gh `
    -v $PWD/config.js:/usr/src/app/config.js `
    -v /var/run/docker.sock:/var/run/docker.sock `
    -v /tmp:/tmp renovate/renovate:slim `
    renovate-tests/gomod1
```
